zeppelin-web-angular/e2e/models/notebook-keyboard-page.ts

```diff
@@ -0,0 +1,674 @@
+  }
+
+  // Switch output show/hide - control.alt.o (or control.alt.ø for macOS)
+  async pressSwitchOutputShow(): Promise<void> {
+    await this.executePlatformShortcut(['control.alt.o', 'control.alt.ø']);
+  }
+
+  // Switch line numbers - control.alt.m (or control.alt.µ for macOS)
+  async pressSwitchLineNumber(): Promise<void> {
+    await this.executePlatformShortcut(['control.alt.m', 'control.alt.µ']);
+  }
+
+  // Switch title show/hide - control.alt.t (or control.alt.† for macOS)
+  async pressSwitchTitleShow(): Promise<void> {
+    await this.executePlatformShortcut(['control.alt.t', 'control.alt.†']);
+  }
+
+  // Clear output - control.alt.l (or control.alt.¬ for macOS)
+  async pressClearOutput(): Promise<void> {
+    await this.executePlatformShortcut(['control.alt.l', 'control.alt.¬']);
+  }
+
+  // Link this paragraph - control.alt.w (or control.alt.∑ for macOS)
+  async pressLinkParagraph(): Promise<void> {
+    await this.executePlatformShortcut(['control.alt.w', 'control.alt.∑']);
+  }
+
+  // Reduce paragraph width - control.shift.-
+  async pressReduceWidth(): Promise<void> {
+    await this.executePlatformShortcut(['control.shift.-', 'control.shift._']);
+  }
+
+  // Increase paragraph width - control.shift.=
+  async pressIncreaseWidth(): Promise<void> {
+    await this.executePlatformShortcut(['control.shift.=', 'control.shift.+']);
+  }
+
+  // Cut line - control.k
+  async pressCutLine(): Promise<void> {
+    await this.executePlatformShortcut('control.k');
+  }
+
+  // Paste line - control.y
+  async pressPasteLine(): Promise<void> {
+    await this.executePlatformShortcut('control.y');
+  }
+
+  // Search inside code - control.s
+  async pressSearchInsideCode(): Promise<void> {
+    await this.executePlatformShortcut('control.s');
+  }
+
+  // Find in code - control.alt.f (or control.alt.ƒ for macOS)
+  async pressFindInCode(): Promise<void> {
+    await this.executePlatformShortcut(['control.alt.f', 'control.alt.ƒ']);
+  }
+
+  async getParagraphCount(): Promise<number> {
+    if (this.page.isClosed()) {
+      return 0;
+    }
+    return await this.paragraphContainer.count();
+  }
+
+  getParagraphByIndex(index: number): Locator {
+    return this.paragraphContainer.nth(index);
+  }
+
+  async isAutocompleteVisible(): Promise<boolean> {
+    return await this.autocompletePopup.isVisible();
+  }
+
+  async getAutocompleteItemCount(): Promise<number> {
+    if (await this.isAutocompleteVisible()) {
+      return await this.autocompleteItems.count();
+    }
+    return 0;
+  }
+
+  async isParagraphRunning(paragraphIndex: number = 0): Promise<boolean> {
+    const paragraph = this.getParagraphByIndex(paragraphIndex);
+    const runningIndicator = paragraph.locator('.paragraph-control .fa-spin, .running-indicator');
+    return await runningIndicator.isVisible();
+  }
+
+  async hasParagraphResult(paragraphIndex: number = 0): Promise<boolean> {
```

Comment:
The method name might be misread as checking only the existence of the paragraph result element.
Since the implementation actually checks whether the result’s status is in a determinable state, maybe a clearer name would be something like `isParagraphResultSettled`.